Parameter Sweep for Attention

dorde-antic opened the pull request:
Implement parameter sweep for attention (`attentionSweeps.py`) that will test combinations of input shapes and perfConfigs for attention and find potential bugs. Adjust `parameterSweeps.py` so that attentionSweeps.py can reuse its methods.

Resolves https://github.com/ROCm/rocMLIR-internal/issues/1800

dorde-antic committed:
Initial commit

dorde-antic committed:
Implement testAttentionConfig and use AttentionConfiguration from perfRunner instead of inheriting it.

dorde-antic committed:
Add values range for seq_lengths, head_dims. Make cocode cleaner

dorde-antic committed:
Add perfConfigSpace

dorde-antic committed:
Describe methods

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,212 @@
+        DATA_TYPES_ATTENTION, GROUPS, SEQ_LENGTHS, SEQ_LENGTHS, HEAD_DIMS, HEAD_DIMS, BOOLS, BOOLS, BOOLS, BOOLS, BOOLS
+    ))
+
+    perfConfigSpace = list(itertools.product(
```
this is not the attention perf config. See Rock_AttnPerfConfig in RockAttrDefs.td or RockTuningImpl.cpp (createAttnTuningRangeBF)

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,212 @@
+
+# GLOBAL VARIABLES
+DATA_TYPES_ATTENTION = ['i8', 'f32', 'f16', 'bf16']
+SEQ_LENGTHS = [1, 27, 64, 77, 128, 256, 384, 1024, 2048, 4097, 8192, 16384]
```
I'd prefer to sample from 1 to 16384, including all possible numbers. Same for all variables.

dhernandez0 replied:
maybe you can represent this as a range() instead of a list

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,212 @@
+DATA_TYPES_ATTENTION = ['i8', 'f32', 'f16', 'bf16']
+SEQ_LENGTHS = [1, 27, 64, 77, 128, 256, 384, 1024, 2048, 4097, 8192, 16384]
+HEAD_DIMS = [32, 64, 96, 128, 512]
+GROUPS = [1, 4, 8]
```
group can be bigger (see attention-configs where it's 256), of course, some combination might not fit into memory. We have to avoid failing if the issue is memory.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,212 @@
+    paths = createPaths(None, args.mlir_build_dir)
+    options = Options(debug=args.debug, quiet=args.quiet, arch=arch, flags=[], concurrent_tests=args.jobs)
+   
+    attentionShapes = list(itertools.product(
```
there are some things not represented here (because of AttentionConfiguration):
- causal: coming in https://github.com/ROCm/rocMLIR/pull/1829)
- with_attn_bias: implemented in rocmlir-gen but for some reason not in AttentionConfiguration
- GQA: this is group query attention, in rocmlir-gen -num_heads_q 4 -num_heads_kv 2, not implemented in AttentionConfiguration
- KV-Cache: -current_seq_len=17,1,32 where the size must match -g 3 (3 numbers for current_seq_len). These are dynamic seq_lens (known at run-time). Not implemented in AttentionConfiguration.

dorde-antic replied:
- causal: waiting until PR is merged
- with_attn_bias: implemented
- GQA: https://github.com/ROCm/rocMLIR-internal/issues/1834
- KV-Cache: https://github.com/ROCm/rocMLIR-internal/issues/1835

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,212 @@
+        [0, 1] # aCopyMore/forceUnroll
+    ))
+    
+    samples = random.sample(list(itertools.product(attentionShapes, perfConfigSpace)), args.samples)
```
for kv-cache, we can probably have another call to sampling. Because seq_len_q = 1, seq_len_kv = maxLenSet (which could be something like 1024, 2048, ...), then we need current_seq_len... 

Maybe we can leave kv-cache for a follow up PR?

dorde-antic committed:
perfConfig for attention fix, value ranges fix

dorde-antic committed:
Value ranges fix

dorde-antic committed:
Code fixes

dorde-antic committed:
Change sampling method to avoid OOM crashes

dorde-antic committed:
Remove unnecessary print() lines

dorde-antic committed:
AttentionConfiguration generateMlirDriverArgs() method for attentionSweeps

dorde-antic committed:
Pipe chaining fixed

dorde-antic committed:
Typo fix

dorde-antic committed:
Fix

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,304 @@
+    parser.add_argument('--quiet', action='store_true')
+    parser.add_argument('--jobs', type=int, default=4)
+    parser.add_argument('--mlir-build-dir', type=str, required=True)
+    parser.add_argument('--samples', type=int, default=20)
```
make default bigger

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,304 @@
+            print(' '.join(fail.generateMlirDriverArgs([])))
+        if args.log_failures:
+            logFailingConfigs(failing, LOGFILE)
+
```
return something because the return value is used below.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,304 @@
+        random.randint(1, 16384), # SEQ_LEN_Q
+        random.randint(1, 16384), # SEQ_LEN_K
+        random.randint(1, 16384), # HEAD_DIM_QK
+        random.randint(1, 16384), # HEAD_DIM_V
```
I don't think head dims are typically that large, maybe 512 or 1024 max?

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,304 @@
+    perfConfig versions.'''
+    V1 = 'v1'
+    V2 = 'v2'
+    V3 = 'v3'
```
there's only v1 for attention

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,304 @@
+from perfRunner import MLIR_N_REPEATS
+
+# GLOBAL VARIABLES
+DATA_TYPES_ATTENTION = ['f32', 'f16', 'bf16']
```
i8 missing

dorde-antic committed:
Fix

dorde-antic committed:
Fixes

dorde-antic committed:
Pipes fix

dorde-antic committed:
Remove unnecessary print()

dorde-antic committed:
Mlir runner command args fix

dorde-antic committed:
IIncreasedefault for samples number, small changes

dorde-antic committed:
IIncreasedefault for samples number, small changes

dorde-antic committed:
Consider hipErrorOutOfMemory as INVALID instead of FAILED

dorde-antic committed:
Comment edit

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,311 @@
+            print("Runner failed:\nOutput = ", stdout3.decode().strip())
+            print("\nError = ", stderr3.decode().strip())
+        
+        if 'hipErrorOutOfMemory' in stderr3.decode().strip():
```
to decide if something fits into memory and generally if it can be run, you can use:
rocmlir-gen ... | rocmlir-driver --kernel-pipeline=applicability

if it fails applicability it means, for some reason we can't run this problem with that particular perfconfig. 
You can check parametersSweep.py, they do the same for convs there.

dhernandez0 replied:
Ideally, we run applicability, if it fails, return TestResult.INVALID. Any other failure would be TestResult.FAIL

dorde-antic commented:
https://github.com/ROCm/rocMLIR-internal/issues/1835#issuecomment-2894645293
_For the randomness, make sure it is re-creatable. So the seed should be capable of rerunning over and over. Possibly use the week of the year so the same random number is used for 7 days_

dorde-antic commented:
**when sampling num-heads (gqa):**
- sample x=[1,n] natural numbers, then do 2**x
- num_heads_q >= num_heads_kv and num_heads_q % num_heads_kv == 0 are requirements

dorde-antic committed:
Generate current_seq_lens randomly with seed based on week number

dorde-antic committed:
Add current_seq_len to mlir command

dorde-antic committed:
Console output edit

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,311 @@
+    if proc1.returncode !=0:
+        if options.debug:
+            print("rocmlir-gen failed:\nError = ", rocmlirGeneratorError.decode().strip())
+        return TestResult.INVALID
```
I think if anything fails should be fail. The only INVALID should come from applicability failing.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,311 @@
+    result += rocmlir_gen_flags or []
+    return result
+
+AttentionConfiguration.generateMlirDriverArgs = generateMlirDriverArgs
```
why do we need to do this?

dorde-antic replied:
To avoid making whole new class that would inherit AttentionConfiguration (like it was done in _parameterSweeps.py_) and to avoid adding new class method in _perfRunner_ inside AttentionConfiguration class which would be only used in _attentionSweeps.py_ file (at least for now). I thought we could extend AttentionConfiguration class dynamically with generateMlirDriverArgs() method which is implemented in _attentionSweeps.py_ file using line 83. Without "attaching" generateMlirDriverArgs like this to AttentionConfiguration, we would get AttributeError, since objects of AttentionConfiguration class wouldn't have attribute generateMlirDriverArgs which we call in testAttentionConfig method.

dhernandez0 replied:
can't we use generateMlirDriverCommandLine that already exists in AttentionConfiguration?

dhernandez0 replied:
that one has causal, which generateMlirDriverArgs is missing. And it seems it's pretty much the same thing?

dorde-antic committed:
Sample num_heads_q and num_heads_kv

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+
+# Week number is used as seed
+seed = datetime.utcnow().isocalendar()[1]
+random.seed(seed)
```
do we use any other library to get random numbers?

dorde-antic replied:
Resolved on Teams (We use built-in module in python.)

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+    perfString = f"attn:v1:{','.join(str(x) for x in perf)}"
+    return AttentionConfiguration(
+        dtype, g, slq, slk, hdqk, hdv, scale,
+        bias, tq, tk, tv, to, options.arch,
```
causal is missing here.

dhernandez0 replied:
I'd prefer to use dtype=dtype, ... so that if at some point we add new args to the constructor this will still work

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+    
+    Requirements:
+        - numHeadsQ >= numHeadsKV
+        - numHeadsQ % numHeadsKV == 0'''
```
nice comment!

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+            numHeadsQ = 2**random.randint(1, 6)
+            numHeadsKV = 2**random.randint(1, 6)
+
+            if numHeadsQ >= numHeadsKV and numHeadsQ%numHeadsKV == 0: # found valid case
```
if it's equal, we are actually not running GQA.

dorde-antic replied:
Yes, but it's the way that we sample other valid values of numHeadsQ and numHeadsKV that are not default (not 1) but equal (for example both of them 2 or 8... ).

dhernandez0 replied:
I'm not sure I follow what you mean. What I mean is, we decided previously whether GQA is going to run (genNumHeads is true). Then, we want GQA to run, so we have to make sure numHeadsQ > numHeadsKV (not ==), otherwise, if they are equal GQA will not run.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+    Requirements:
+        - numHeadsQ >= numHeadsKV
+        - numHeadsQ % numHeadsKV == 0'''
+    genNumHeads = random.choice(BOOLS)
```
we might want non-50% chance here. But let's discuss that in the next meeting.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+def sampleAttentionShape():
+    global CURRENT_SEQ_LEN
+
+    g = random.randint(1, 16384)
```
let's do random.randint(1, 256) here

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+        random.randint(1, 1024), # HEAD_DIM_QK
+        random.randint(1, 1024), # HEAD_DIM_V
+        random.choice(BOOLS),   # with_attn_scale
+        random.choice(BOOLS),   # with_attn_bias
```
missing causal

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+        [32, 64, 128, 256], # M/Wave
+        [4, 16, 32], # MN/Xdl
+        [4, 8, 16], # kPack
+        [0, 1] # forceUnroll
```
due to an old bug (nobody got around to fix it, maybe it doesn't happen anymore?), forceUnroll is always 1. https://github.com/ROCm/rocMLIR-internal/issues/790.

No need to iterate over 0, 1

dhernandez0 replied:
On second thought, let's leave it like this. So, when/if it's fixed, it's already iterating over it. Currently, 0 will fail on applicability really fast.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+    )
+
+
+perfConfigSpace = list(itertools.product(
```
note that the perfConfig is different for wmma/mfma. The code here should take that into account.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+    seqLenK = random.randint(1, 16384) # SEQ_LEN_K 
+
+    useKVCache = random.choice(BOOLS)
+    CURRENT_SEQ_LEN = genCurrentSeqLens(g, seqLenK) if useKVCache else None
```
if useKVCache is true, we should force seqLenQ = 1. To emulate what kv-cache looks like. Also, maybe make genNumHeads more likely (higher probability).

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+GFX_CHIP_RE = re.compile(r"gfx[0-9a-z]+")
+CURRENT_SEQ_LEN = None
+
+# Week number is used as seed
```
expand comment: to make sure weekly CI is reproducible... something like that

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+        f"-transK={self.transK}",
+        f"-transV={self.transV}",
+        f"-transO={self.transO}",
+        '--kernel-repeats', str(MLIR_N_REPEATS),
```
in this case, as we don't want to measure run-time, 1 is ok. It'd be faster as well.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+            description='Sweep parameter values for attention to detect bugs')
+    parser.add_argument('--debug', action='store_true')
+    parser.add_argument('--quiet', action='store_true')
+    parser.add_argument('--jobs', type=int, default=4)
```
is it possible to default to num cpus/cores somehow? if it's not easy let's leave it like this.

dorde-antic replied:
The easiest way would be using default=os.cpu_count() if that is okay? Tried on Navi4x, returns 32

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+
+
+def genCurrentSeqLens(g: int, maxSeqLen: int) -> list[int]:
+    return [random.randint(1, maxSeqLen) for _ in range(g)]
```
due to a recent change: https://github.com/ROCm/rocMLIR/pull/1827. The valid range is 0 to maxSeqLen-1 (both included).

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,359 @@
+
+
+def sampleAttentionShape():
+    global CURRENT_SEQ_LEN
```
can we reutrn this as the other params instead of using a global?

dorde-antic commented:
TODO:
- [x]  Consider the new flag from https://github.com/ROCm/rocMLIR/pull/1882

dorde-antic committed:
Add causal and return_lse flags, small fixes

dorde-antic committed:
Merge branch 'develop' into attentionSweeps

dorde-antic committed:
Change valid range for getCurrentSeqLens()

dorde-antic committed:
Addressing comments

dorde-antic committed:
Addressing comments

dorde-antic committed:
Don't use global CURRENT_SEQ_LEN

dorde-antic committed:
Set --kernel-repeats to 1

dorde-antic committed:
Reuse generateMlirDriverCommandLine from perfRunner

dorde-antic committed:
Remove unnecessary debug message

dorde-antic committed:
Merge branch 'develop' into attentionSweeps

dorde-antic committed:
List to string: new changes

dorde-antic committed:
Cleaner repr in log

dorde-antic committed:
Cleaner repr in log - missing comma when split

dorde-antic committed:
Consider two types of perfConfigs

dorde-antic committed:
newline

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+Options:
+    --mlir-build-dir    Path to the MLIR build directory (required)
+    --samples           Number of random configuration samples to the test (default: 1000)
+    --jobs              Number of concurrent tests to run in parallel (default: 4)
```
default: os.cpu_count()

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    parser.add_argument('--debug', action='store_true')
+    parser.add_argument('--quiet', action='store_true')
+    parser.add_argument('--jobs', type=int, default=os.cpu_count())
+    parser.add_argument('--mlir-build-dir', type=str, required=True)
```
use perfRunner.find_mlir_build_dir() to have a default value, see parameterSweeps.py. Then, make it not required.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    parser.add_argument('--quiet', action='store_true')
+    parser.add_argument('--jobs', type=int, default=os.cpu_count())
+    parser.add_argument('--mlir-build-dir', type=str, required=True)
+    parser.add_argument('--samples', type=int, default=1000)
```
I wonder if we should use a time limit instead of number of samples? Because different machines will take diff time for 1k samples. @umangyadav

dorde-antic replied:
> I wonder if we should use a time limit instead of number of samples? Because different machines will take diff time for 1k samples. @umangyadav @dhernandez0 

There is option to add timeout in the CI function which would call attentionSweeps.py, which can be used independently of the samples number.

dorde-antic replied:
Maybe void parameterSweep(...) in CI, could be changed to something like:
```
void parameterSweep(String CONFIG, String codepath, String sweepType = "parameter") {
    timeout(time: 300, activity: true, unit: 'MINUTES') {
        dir('build') {
            if (sweepType == "attention") {
                sh """python3 ./bin/attentionSweeps.py"""
            } else {
                sh """python3 ./bin/parameterSweeps.py -j 5 ${CONFIG}"""
            }
        }
    }
}
```
it should be time limited then I think ...

and call in the stage would be something like:
```
stage("Parameter Sweep") {
                        steps {
                            script {
                         // ...
                                    ]) {
                                        parameterSweep("conv_structure", "${CODEPATH}")
                                        parameterSweep("perf_config", "${CODEPATH}")
                                        parameterSweep("", "${CODEPATH}", "attention")
                                    }
                                }
                            }
                        }       
                    }
```

dhernandez0 replied:
but I think if it times out, it will make CI fail?

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+from perfRunner import create_paths as createPaths
+
+# GLOBAL VARIABLES
+DATA_TYPES_ATTENTION = ['i8', 'f32', 'f16', 'bf16']
```
import from perfRunner

dhernandez0 replied:
also, wmma doesn't support f32. We need to have some kind of special case for that.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+DATA_TYPES_ATTENTION = ['i8', 'f32', 'f16', 'bf16']
+BOOLS = [True, False]
+LOGFILE = 'failing_configs.csv'
+GFX_CHIP_RE = re.compile(r"gfx[0-9a-z]+")
```
import from perfRunner

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+seed = datetime.utcnow().isocalendar()[1]
+random.seed(seed)
+
+class TestResult(Enum):
```
Import from paramterSweep.py, it's already defined there.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    FAIL = 3
+    
+@dataclass(frozen=True)
+class Options:
```
import from perfRunner

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    return attnConfig
+
+
+def multilineRepr(obj, num_fields=4):
```
can we use textwrap:
import textwrap

wrapped_text = textwrap.fill(text, width)

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    return '\n'.join(lines)
+
+
+async def testAttentionConfig(config: AttentionConfiguration, options: Options, paths) -> TestResult:
```
you are not using kernel-pipeline=applicability to decide if it's TestResult.INVALID. All the rest of failures must be TestResult.FAIL

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+
+
+IterType = TypeVar('IterType')
+def grouper(iterable: Iterable[IterType], n: int):
```
import from parameterSweep.py

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    """Test the given `params`, returning the corresponding `config` on failure
+    and `None` on success or inapplicability."""
+    result = await testAttentionConfig(config, options, paths)
+
```
see paramterSweep dropGoodConfig function, it prints out value as well:
if not options.quiet:
        print(f"{result.name}: {config!r}")

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+        yield chunk
+
+
+async def dropGoodConfig(config: AttentionConfiguration,
```
can we use the function in parameterSweep.py? maybe pass the function instead of calling testAttentionConfig or testConfig

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    return config if result == TestResult.FAIL else result
+
+
+async def sweepParameters(paramIter: Iterable[IterType],
```
same here, try to use the function in paramterSweep, passing functions if needed.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    )
+
+
+perfConfigSpaceMFMA = list(itertools.product( # MFMA perfConfig space
```
add a comment to keep this in sync with RockTuningImpl.cpp. Also, please add a comment in RockTuningImpl.cpp as well to keep in sync with this file.

dhernandez0 commented on mlir/utils/performance/attentionSweeps.py:
```diff
@@ -0,0 +1,417 @@
+    if not args.quiet:
+        print(f"Sampling {args.samples} configurations from attention space...")
+
+    if chip in ['gfx900', 'gfx906', 'gfx908', 'gfx90a', 'gfx940', 'gfx941', 'gfx942']:
```
add gfx950 as well, gfx940 and gx941 are deprecated. Maybe, we can even do .startsWith('gfx9')?

dhernandez0 replied:
also, add a TODO to use AmdArchDb python version when available.

dorde-antic committed:
Address comments

dorde-antic committed:
Address comments

Signed-off-by: Djordje Antic <djoantic@amd.com>

dorde-antic committed:
Merge branch 'develop' into attentionSweeps

dorde-antic committed:
Reuse methods

Signed-off-by: Djordje Antic <djoantic@amd.com>

dorde-antic committed:
Refactor code

Signed-off-by: Djordje Antic <djoantic@amd.com>

dhernandez0 reviewed (approved):
LGTM, thanks for creating this script, great work!

dorde-antic committed:
Merge branch 'develop' into attentionSweeps

dorde-antic commented:
TODO (in a follow-up tickets/PRs):
- Consider the new flag from https://github.com/ROCm/rocMLIR/pull/1895 when it's merged
- Add attentionSweeps.py to CI in https://github.com/ROCm/rocMLIR-internal/issues/1837

dorde-antic committed:
Adjust attentionSweeps and perfRunner with the DATA_TYPES_ATTENTION changes
Add return value to initializeDataTypesAttention() in perfRunner so that it can be used easily in other modules.
Invoke initializeDataTypesAttention() in attentionSweeps.py.

Signed-off-by: Djordje Antic <djoantic@amd.com>

dorde-antic commented:
`perfRunner.py` and `tuningRunner.py` were tested after change in perfRunner from [67f0f4c](https://github.com/ROCm/rocMLIR/pull/1830/commits/67f0f4c84da1021041d5f5c2340ca0efece6d367)
